src/jupyterlabcontroller/services/prepuller.py

```diff
@@ -0,0 +1,883 @@
+                del self._schedulers[image]
+        if "main" in self._schedulers and self._schedulers["main"] is not None:
+            self.logger.warning(
+                "Terminating main prepuller executor scheduler"
+            )
+            await self._schedulers["main"].close()
+            del self._schedulers["main"]
+
+    async def create_prepuller_pod_spec(
+        self,
+        image: str,
+        node: str,
+    ) -> PodSpec:
+        shortname = image.split("/")[-1]
+
+        #        user = UserInfo(
+        #            username="prepuller",
+        #            name="Prepuller User",
+        #            uid=1000,
+        #            gid=1000,
+        #            groups=[
+        #                UserGroup(
+        #                    name="prepuller",
+        #                    id=1000,
+        #                )
+        #            ],
+        #        )
+        return PodSpec(
+            containers=[
+                Container(
+                    name=f"prepull-{shortname}",
+                    command=["/bin/sleep", "5"],
+                    image=image,
+                    working_dir="/tmp",
+                )
+            ],
+            node_name=node,
+        )
+
+    async def prepull_images(self) -> None:
+        """This is the method to identify everything that needs pulling, and
+        spawns pods with those images on the nodes that need them.
+        """
+
+        status = await self.get_prepulls()
+
+        pending = status.images.pending
+
+        required_pulls: Dict[str, List[str]] = dict()
+        for img in pending:
+            if img.missing is not None:
+                for i in img.missing:
+                    if i.eligible:
+                        if img.path not in required_pulls:
+                            required_pulls[img.path] = list()
+                        required_pulls[img.path].append(i.name)
+        self.logger.debug(f"Required pulls by node: {required_pulls}")
+        timeout = PREPULLER_PULL_TIMEOUT
+        # Parallelize across nodes but not across images
+        for image in required_pulls:
+            if image in self._schedulers:
+                self.logger.warning(
+                    f"Scheduler for image {image} already exists.  Presuming "
+                    "earlier pull still in progress."
+                )
+                continue
+            scheduler = Scheduler(close_timeout=timeout)
+            self._schedulers[image] = scheduler
+            tag = image.split(":")[1]
+            for node in required_pulls[image]:
+                await scheduler.spawn(
+                    self.k8s_client.create_pod(
+                        name=f"prepull-{tag}",
+                        namespace=self.namespace,
+                        pod=await self.create_prepuller_pod_spec(
+                            image=image,
+                            node=node,
+                        ),
+                    )
+                )
```

Comment:
I'm not sure it does.  It's somewhat changed in the refactor, but the idea is each image is done serially (and more or less does go in that order, because of the ordering of the RSPTagType enum), but spawns a bunch of pods across nodes in parallel.  I think I got that right?